File: pythonX/medium/m016_number_of_islands.py
"""
M016: Number of Islands

Summary:
Count number of islands in 2D grid (DFS/BFS).

Requirements:
Given grid of '1's (land) and '0's (water), count islands.
Islands formed by connecting adjacent lands horizontally/vertically.

Approach:
Use DFS to mark visited land cells.

Data Structures:
- 2D grid (list[list[str]])

Big O Analysis:
Time Complexity: O(m * n)
Space Complexity: O(m * n) for recursion stack
"""


def num_islands(grid: list[list[str]]) -> int:
    """
    TODO: Count islands
    1. If grid empty, return 0
    2. Initialize count = 0
    3. Loop through each cell:
       - If grid[i][j] == '1':
         - Increment count
         - Call dfs to mark entire island
    4. Return count
    """
    pass


def dfs(grid: list[list[str]], i: int, j: int) -> None:
    """
    TODO: DFS to mark island
    1. Check boundaries
    2. If not '1', return
    3. Mark as visited (set to '0')
    4. Recursively visit 4 neighbors
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Single large island")
    grid1 = [
        ['1','1','1','1','0'],
        ['1','1','0','1','0'],
        ['1','1','0','0','0'],
        ['0','0','0','0','0']
    ]
    print(f"Number of islands: {num_islands(grid1)}")  # Expected: 1
    
    print("\nTest 2: Multiple islands")
    grid2 = [
        ['1','1','0','0','0'],
        ['1','1','0','0','0'],
        ['0','0','1','0','0'],
        ['0','0','0','1','1']
    ]
    print(f"Number of islands: {num_islands(grid2)}")  # Expected: 3
    
    print("\nTest 3: Checkerboard pattern")
    grid3 = [
        ['1','0','1','0','1'],
        ['0','1','0','1','0'],
        ['1','0','1','0','1']
    ]
    print(f"Number of islands: {num_islands(grid3)}")  # Expected: 9

File: src/com/example/javas/medium/M016_NumberOfIslands.java
/*
MEDIUM PROBLEM (Simplified for 25 min):
Number of Islands (Graph/Matrix DFS)

REQUIREMENTS:
- Given a 2D grid of '1's (land) and '0's (water)
- Count the number of islands
- An island is surrounded by water and formed by connecting adjacent lands horizontally or vertically

APPROACH:
- Use DFS to mark visited land cells
- Iterate through grid, when '1' is found, increment island count and DFS to mark all connected land

TIME COMPLEXITY: O(m * n) where m, n are grid dimensions
SPACE COMPLEXITY: O(m * n) for recursion stack in worst case
*/

package com.example.javas.medium;

public class M016_NumberOfIslands {
    
    public static int numIslands(char[][] grid) {
        // TODO: Implement island counter
        // 1. If grid is empty, return 0
        // 2. Initialize count = 0
        // 3. Loop through each cell (i, j):
        //    - If grid[i][j] == '1':
        //      - Increment count
        //      - Call dfs(grid, i, j) to mark entire island
        // 4. Return count
        
        return 0;
    }
    
    private static void dfs(char[][] grid, int i, int j) {
        // TODO: Implement DFS to mark island
        // 1. Check boundaries: if i or j out of bounds, return
        // 2. If grid[i][j] != '1', return (water or already visited)
        // 3. Mark current cell as visited: grid[i][j] = '0'
        // 4. Recursively visit 4 neighbors:
        //    - dfs(grid, i+1, j) // down
        //    - dfs(grid, i-1, j) // up
        //    - dfs(grid, i, j+1) // right
        //    - dfs(grid, i, j-1) // left
        
    }
    
    // TEST CASES
    public static void main(String[] args) {
        char[][] grid1 = {
            {'1','1','1','1','0'},
            {'1','1','0','1','0'},
            {'1','1','0','0','0'},
            {'0','0','0','0','0'}
        };
        System.out.println("Number of islands in grid1: " + numIslands(grid1));
        
        char[][] grid2 = {
            {'1','1','0','0','0'},
            {'1','1','0','0','0'},
            {'0','0','1','0','0'},
            {'0','0','0','1','1'}
        };
        System.out.println("Number of islands in grid2: " + numIslands(grid2));
        
        char[][] grid3 = {
            {'1','0','1','0','1'},
            {'0','1','0','1','0'},
            {'1','0','1','0','1'}
        };
        System.out.println("Number of islands in grid3: " + numIslands(grid3));
    }
}
